`cached` cache population daemon

airhorns opened the pull request:
This adds a new binary that implements a caching daemon -- a long running agent that sits beside DL workloads that's able to very quickly populate a cache for those workloads. I called it `cached` ("cache-dee") in the sense that it is a daemon that caches. We intend to run this daemon in k8s as a daemonset, but that's in the next PR. 

The fundamental approach with this thing is to hardlink the cache from a shared spot into each cache user's working directory. The DL client running in the cache user's workspace will then hardlink what it can from the cache, and download whatever else it needs.

More explicitly:
  - we'll have a golden copy of the cache on the k8s host at /var/lib/kubernetes/dateilager-cache (has to be here to be on the same device)
  - we'll then hardlink from there into /var/lib/kubernetes/pods/volumes/kubernetes.io~empty-volume/gadget/cache, which will appear as `/gadget/cache` within the container
  - the DL client within the container will then build the app's FS at `/gadget/app`, using `/gadget/cache` as the cache, hardlinking a second time from there.

Since hardlinks share content, this means that the cache is shared on disk between all the users, and we dont need to do an expensive file copy or take up more space. The downside is that they are actually the same files in each place they exist, so if one entry gets modified, the other would too. This is very bad, as it'd allow one sandboxed cache user to affect the other sandboxed users view of the cache. So, we rely on UNIX permissions to ensure that each cache user can't modify the shared cache in prod.

So, in service of running this sucker as a daemonset in CI, this introduces: 
 - a new binary, in the same dockerfile, called `cached`
 - a new GRPC proto definition that that binary serves, with really only one command, which makes the cached server place the cache in a path

This lets us mess around with it locally  and sets the stage for #89 where we k8s-ify this thing.

You can 🎩 if you want with
 - checkout branch, start dev, start server with `make server`
 - start `cached` with `make cached`
 - run a `make client-getcache-from-daemon` to have the client fetch the cache via the daemon
 - run a `make client-rebuild-with-cache` to rebuild using the cache

scott-rc commented:
🎩'd this locally with Gadget. Used the editor, filesync, ran some tests and all looked good.

cc. @angelini

airhorns commented:
* **#91** <a href="https://app.graphite.dev/github/pr/gadget-inc/dateilager/91?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#89** <a href="https://app.graphite.dev/github/pr/gadget-inc/dateilager/89?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#85** <a href="https://app.graphite.dev/github/pr/gadget-inc/dateilager/85?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#90** <a href="https://app.graphite.dev/github/pr/gadget-inc/dateilager/90?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#86** <a href="https://app.graphite.dev/github/pr/gadget-inc/dateilager/86?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#87** <a href="https://app.graphite.dev/github/pr/gadget-inc/dateilager/87?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @airhorns and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

airhorns committed:
Add a "shared reader" auth mode for accessing caches without a project or admin token

airhorns committed:
Implement a cache daemon that can move the cache into place quickly on demand

 - a long running background process we'll deploy as a daemonset on k8s nodes
 - moves the cache into place via hardlinking the golden copy from a shared location
 - operates using the host filesystem, and thus has pretty priviledged access

airhorns commented on Makefile:
```diff
@@ -8,11 +8,13 @@ DB_USER ?= postgres
-DEV_TOKEN_PROJECT_1 ?= v2.public.eyJzdWIiOiIxIiwiaWF0IjoiMjAyMS0xMC0xNVQxMToyMDowMC4wMzVaIn2MQ14RfIGpoEycCuvRu9J3CZp6PppUXf5l5w8uKKydN3C31z6f6GgOEPNcnwODqBnX7Pjarpz4i2uzWEqLgQYD
+DEV_TOKEN_ADMIN ?= v2.public.eyJzdWIiOiJhZG1pbiJ9yt40HNkcyOUtDeFa_WPS6vi0WiE4zWngDGJLh17TuYvssTudCbOdQEkVDRD-mSNTXLgSRDXUkO-AaEr4ZLO4BQ
+DEV_TOKEN_PROJECT_1 ?= v2.public.eyJzdWIiOiIxIn2jV7FOdEXafKDtAnVyDgI4fmIbqU7C1iuhKiL0lDnG1Z5-j6_ObNDd75sZvLZ159-X98_mP4qvwzui0w8pjt8F
+DEV_SHARED_READER_TOKEN ?= v2.public.eyJzdWIiOiJzaGFyZWQtcmVhZGVyIn1CxWdB02s9el0Wt7qReARZ-7JtIb4Zj3D4Oiji1yXHqj0orkpbcVlswVUiekECJC16d1NrHwD2FWSwRORZn8gK
```
had to re-sign the tokens cause I couldn't find the private key, included a new keypair in the repo now :+1:

airhorns commented on internal/pb/cache.proto:
```diff
@@ -0,0 +1,14 @@
+
+option go_package = "github.com/gadget-inc/dateilager/pkg/pb";
+
+service Cache {
```
This GRPC service is *not* what we'll use in production -- it's just for testing, but it lets us ask the daemon to do its thing in tests and in development, very convenient.

angelini replied:
In Prod we're going to implement a CSI proto and that's what will be called directly by K8S?

airhorns replied:
Yup!

airhorns commented on internal/testutil/context.go:
```diff
@@ -34,6 +34,9 @@ func NewTestCtx(t *testing.T, role auth.Role, projects ...int64) TestCtx {
 	})
 
+	log := zaptest.NewLogger(t)
+	zap.ReplaceGlobals(log)
```
This makes a zap logger that will log in tests for debugging, but because it's a `zaptest.Logger`, it doesn't log unless you run `go test -v`. Me likey.

airhorns commented on pkg/api/fs.go:
```diff
@@ -948,7 +966,7 @@ func (f *Fs) GetCache(req *pb.GetCacheRequest, stream pb.Fs_GetCacheServer) erro
 	trace.SpanFromContext(ctx)
 
-	_, err := requireProjectAuth(ctx)
+	err := requireSharedReaderAuth(ctx)
```
I didn't want `cached` to have an admin token or any specific project's token as its the only DL component that actually runs on k8s nodes next to sandboxed workloads. So, I added a new token type that can read the cache and that's it.

airhorns commented on pkg/cli/cachedaemon.go:
```diff
@@ -0,0 +1,169 @@
+	flags.StringVar(&headlessHost, "headless-host", "", "Alternative headless hostname to use for round robin connections")
+	flags.UintVar(&timeout, "timeout", 0, "GRPC client timeout (ms)")
+	flags.IntVar(&port, "port", 5053, "cache API port")
+	flags.StringVar(&stagingPath, "staging-path", "", "path for staging downloaded caches")
```
This is where `cached` will put the downloaded cache, and what has to be on the same device as the paths it then later links the cache to.

airhorns commented on Makefile:
```diff
@@ -197,15 +215,14 @@ else
 
 build-local-container:
-	docker build -t dl-local:latest .
+	docker build --load -t dl-local:dev .
```
kubernetes will always pull an image with the `latest` tag, which we don't want to do, cause it can't be pulled from anywhere. we want it to just sit in the local image cache, so we name it something else

airhorns commented on Makefile:
```diff
@@ -244,7 +261,9 @@ else
-	cd js && mkdir -p ./src/pb && npx protoc --experimental_allow_proto3_optional --ts_out ./src/pb --ts_opt long_type_bigint,ts_nocheck,eslint_disable,add_pb_suffix --proto_path ../internal/pb/ ../$^
+	cd js && mkdir -p ./src/pb && for proto in $^; do \
+		npx protoc --experimental_allow_proto3_optional --ts_out ./src/pb --ts_opt long_type_bigint,ts_nocheck,eslint_disable,add_pb_suffix --proto_path ../internal/pb/ ../$$proto; \
+	done
```
multiple protofiles broke `$^`

airhorns commented on flake.nix:
```diff
@@ -72,6 +73,7 @@
               flake.packages.postgresql
               flake.packages.dev
               flake.packages.clean
+              flake.packages.glibcLocales
```
This makes the flake work ok on linux

airhorns commented on Dockerfile:
```diff
@@ -8,6 +8,10 @@ EOF
+RUN GRPC_HEALTH_PROBE_VERSION=v0.4.23 \
+    && curl -Lfso /bin/grpc_health_probe https://github.com/grpc-ecosystem/grpc-health-probe/releases/download/${GRPC_HEALTH_PROBE_VERSION}/grpc_health_probe-linux-${TARGETARCH} \
+    && chmod +x /bin/grpc_health_probe
+
```
moved this up so we don't need curl in the container below

angelini replied:
Fair enough, but considering the size of `curl` I've often found it handy when debugging networking problems in prod. I'm not sure I'd want to give that up.

We will also be able to fully drop this after our next K8S upgrade, they built this into core K8S as of version 1.27.

airhorns replied:
This is no longer a thing

angelini commented on js/src/pb/cache_pb.client.ts:
```diff
@@ -0,0 +1,39 @@
+/* eslint-disable */
```
Do we need JS clients for this? Or is that just happening because of the Makefile pattern?

airhorns replied:
The latter -- I figured it was simpler to just make it work so the next person doesn't have to figure out why it isn't being generated, but I could hardcode the js client to just build the FS one

airhorns replied:
I take it back, removing these

angelini commented on pkg/api/cached.go:
```diff
@@ -0,0 +1,99 @@
+}
+
+func (c *Cached) PopulateDiskCache(ctx context.Context, req *pb.PopulateDiskCacheRequest) (*pb.PopulateDiskCacheResponse, error) {
+	err := requireAdminAuth(ctx)
```
Shouldn't this require shared reader?

airhorns replied:
Because this actually modifies the filesystem at an arbitrary path using the permissions of the daemon, and because we don't actually want to use this for much other than testing locally, I figured better to make it restrictive and accept admin only. I think if we ever have a non-CSI use case for this we can lower it? Or, I can just delete this whole piece too

angelini replied:
We use this pattern for dev only routes (especially for things like resetting a DB) and it's probably a good call here: https://github.com/gadget-inc/dateilager/blob/main/pkg/api/fs.go#L783-L785

angelini commented on pkg/api/cached.go:
```diff
@@ -0,0 +1,99 @@
+	return c.currentVersion, nil
+}
+
+/** Fetch the cache into a spot in the staging dir */
```
Nit: you are changing comments styles within the same. There isn't a strong pattern of public function comments in the repo but Go recommends `//`: https://go.dev/doc/comment

angelini commented on pkg/api/cached.go:
```diff
@@ -0,0 +1,99 @@
+func (c *Cached) Prepare(ctx context.Context) error {
+	start := time.Now()
+	newDir := path.Join(c.StagingPath, randomString())
+	version, count, err := c.Client.GetCache(ctx, newDir)
```
In theory the caching system can support being updated (pulling in a new cache version alongside an existing one) and we could do that to a node after it's been booted up.

But considering the speed of change of caches so far, I think a one shot is fine and new cache versions will appear as nodes cycle in and out.

In the future we likely want to expose some kind of GRPC endpoint to force every node to update it's cache to the latest version.

airhorns replied:
Yep definitely -- the `Prepare` function is built to do that, where it uses a new subdir to stage the cache in over time so it will support the nice pattern of preparing a new one while still serving the old one I think that's an easy improvement but I tried to be disciplined and just get something working for now.

angelini replied:
So they are actually safe to build within the same directory.

The objects don't overlap (and if they do, DL should be smart enough to not write a hash that already exists) and then once all of the objects are written the version file is updated to contain a new row with the new version.

So you could fold multiple caches into one directory and it's race safe. That's why `availableCacheVersions` is a list.

airhorns replied:
I think it might make sense to still keep the caches in different directories so that we don't have an ever-growing list of files to hardlink into place for each specialization -- cool to leave as is?

angelini replied:
Yeah I think we'll come back to the updating cache later, we've got a few bridges to cross before nailing that down.

angelini commented on pkg/api/cached.go:
```diff
@@ -0,0 +1,99 @@
+	// Generate a secure random string for the temporary directory name
+	randBytes := make([]byte, 10) // Adjust the size of the byte slice as needed
+	if _, err := rand.Read(randBytes); err != nil {
+		panic(err) // Handle error appropriately in production code
```
Hrmm, this looks like some copy pasta.

angelini commented on test/shared_test.go:
```diff
@@ -646,6 +651,23 @@ func verifyTarResults(t *testing.T, results [][]byte, expected map[string]expect
 
 // Use debugProjects(tc) and debugObjects(tc) within a failing test to log the state of the DB
```
This comment will need to live below `testCert`

angelini commented on test/cached_test.go:
```diff
@@ -0,0 +1,107 @@
+	tmpDir := emptyTmpDir(t)
+	defer os.RemoveAll(tmpDir)
+
+	s := cached.NewServer(tc.Context(), c, testCert(), path.Join(tmpDir, "staging"), pasetoKey)
```
For testing a GRPC server we use the GRPC test tool `buffcon` to actually write over a local connection instead of booting up a real server: https://github.com/gadget-inc/dateilager/blob/main/test/shared_test.go#L371-L385

airhorns replied:
Word, updated to do the same

angelini commented on pkg/cli/cachedaemon.go:
```diff
@@ -0,0 +1,176 @@
+			if err != nil {
+				return err
+			}
+			ctx = client.IntoContext(ctx, cl)
```
Probably not necessary in this context since you don't have child commands that need to pull out the client, you are using it directly a few lines below.

airhorns replied:
👍

angelini commented:
FYI, I'm going to push commits to this branch

angelini committed:
Merge pull request #90 from gadget-inc/cached-prep

Small changes to prep for caching daemon

angelini commented on pkg/client/client.go:
```diff
@@ -46,12 +46,13 @@ type VersionRange struct {
 
 func NewClientConn(conn *grpc.ClientConn) *Client {
-	return &Client{conn: conn, fs: pb.NewFsClient(conn)}
+	return &Client{conn: conn, fs: pb.NewFsClient(conn), cache: pb.NewCacheClient(conn)}
```
As I was playing around with this code this ended up being a bit weird.

We are using a single GRPC connection to talk both to the DL server and to the Cached server.

But in reality if you create a `NewClientConn` with a GRPC connection the the server, the cached client commands won't work and vice versa.

I'm going to refactor this a bit.

angelini committed:
Split clients and consistency changes
